tensorboard/webapp/widgets/data_table/data_table_component.ts

```diff
@@ -202,4 +239,71 @@ export class DataTableComponent implements OnDestroy, AfterContentInit {
   deleteButtonClicked(header: ColumnHeader) {
     this.removeColumn.emit(header);
   }
+
+  focusColumnSelector() {
+    this.columnSelector.focus();
+  }
+
+  openContextMenu(header: ColumnHeader, event: MouseEvent) {
+    event.stopPropagation();
+    event.preventDefault();
+
+    this.contextMenuHeader = header;
+    this.contextMenu.openAtPosition({
+      x: event.clientX,
+      y: event.clientY,
+    });
+  }
+
+  openColumnSelector(event: MouseEvent, insertTo?: Side) {
+    event.stopPropagation();
+    this.insertColumnTo = insertTo;
+    const rect = (event.target as HTMLElement).getBoundingClientRect();
+    this.columnSelectorModal.openAtPosition({
+      x: rect.x + rect.width,
+      y: rect.y + rect.height,
+    });
+  }
+
+  onColumnSelectorClosed() {
+    this.contextMenuHeader = undefined;
+    this.insertColumnTo = undefined;
+  }
+
+  canContextMenuRemoveColumn() {
+    return (
+      this.contextMenuHeader !== undefined &&
+      this.contextMenuHeader.type === ColumnHeaderType.HPARAM
```

Comment:
I do not think the DataTable should be in charge of deciding that only HPARAMs can be removed. Other columns can be removed in the ScalarTable. I like your idea of adding a this sort of metadata to the ColumnHeader interface. 

This may be to much of a refactor for now. However, I think we should at least add a TODO here to remove this and rely on something more generic.

Reply:
+1 for the refactor. I missed this dependency when moving the code to the data table component. I would actually like to see this handled as an optional property added to column header type.